node/network/availability-distribution/src/lib.rs

```diff
@@ -157,164 +154,116 @@ struct PerCandidate {
-			.map(|receipt_set| receipt_set.into_iter())
+			.filter_map(|r| self.per_relay_parent.get(r))
+			.map(|per_relay_parent| per_relay_parent.live_candidates.iter().cloned())
 			.flatten()
-			.map(|(receipt_hash, receipt)| (receipt_hash.clone(), receipt.clone()))
 			.collect()
 	}
 
-	#[tracing::instrument(level = "trace", skip(ctx), fields(subsystem = LOG_TARGET))]
-	async fn add_relay_parent<Context>(
+	#[tracing::instrument(level = "trace", skip(candidates), fields(subsystem = LOG_TARGET))]
+	fn add_relay_parent(
 		&mut self,
-		ctx: &mut Context,
 		relay_parent: Hash,
 		validators: Vec<ValidatorId>,
 		validator_index: Option<ValidatorIndex>,
-	) -> Result<()>
-	where
-		Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
-	{
-		let candidates = query_live_candidates(ctx, self, std::iter::once(relay_parent)).await?;
+		candidates: HashMap<CandidateHash, FetchedLiveCandidate>,
+		ancestors: Vec<Hash>,
+	) {
+		let candidate_hashes: Vec<_> = candidates.keys().cloned().collect();
 
 		// register the relation of relay_parent to candidate..
-		// ..and the reverse association.
-		for (relay_parent_or_ancestor, (receipt_hash, receipt)) in candidates.clone() {
-			self.reverse
-				.insert(receipt_hash.clone(), relay_parent_or_ancestor.clone());
-			let per_candidate = self.per_candidate.entry(receipt_hash.clone()).or_default();
-			per_candidate.validator_index = validator_index.clone();
-			per_candidate.validators = validators.clone();
-
-			self.receipts
-				.entry(relay_parent_or_ancestor)
-				.or_default()
-				.insert((receipt_hash, receipt));
-		}
-
-		// collect the ancestors again from the hash map
-		let ancestors = candidates
-			.iter()
-			.filter_map(|(ancestor_or_relay_parent, _receipt)| {
-				if ancestor_or_relay_parent == &relay_parent {
-					None
-				} else {
-					Some(*ancestor_or_relay_parent)
-				}
-			})
-			.collect::<Vec<Hash>>();
-
-		// mark all the ancestors as "needed" by this newly added relay parent
-		for ancestor in ancestors.iter() {
-			self.ancestry
-				.entry(ancestor.clone())
-				.or_default()
-				.insert(relay_parent);
+		for (receipt_hash, fetched) in candidates {
+			let per_candidate = self.per_candidate.entry(receipt_hash).or_default();
+
+			// Cached candidates already have entries and thus don't need this
+			// information to be set.
+			if let FetchedLiveCandidate::Fresh(descriptor) = fetched {
+				per_candidate.validator_index = validator_index.clone();
+				per_candidate.validators = validators.clone();
+				per_candidate.descriptor = descriptor;
+			}
+			per_candidate.live_in.insert(relay_parent);
 		}
 
-		self.per_relay_parent
-			.entry(relay_parent)
-			.or_default()
-			.ancestors = ancestors;
-
-		Ok(())
+		let per_relay_parent = self.per_relay_parent.entry(relay_parent).or_default();
+		per_relay_parent.ancestors = ancestors;
+		per_relay_parent.live_candidates.extend(candidate_hashes);
```

Comment:
tip: you can move this before the for loop and then do the `extend` without intermediate `collect`ing.